### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by sovchenko**
new pull request

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,29 @@
+pipeline{
+  agent any
+  options{
+    //if
```
Comment (toleksyn) [inline Jenkinsfile]: remove

File: src/main/java/page_elements/filter_section/Filter.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import lombok.Setter;
+import org.openqa.selenium.By;
+import page_objects.ListItemsPage;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Getter
+@Setter
+public class Filter {
```
Comment (toleksyn) [inline src/main/java/page_elements/filter_section/Filter.java]: be specific, ProductFilter

File: src/main/java/page_elements/filter_section/Filter.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    private String filterTitle;
+    private ElementsCollection filterOptions;
+
+    public void selectFilterOption(String filterOptionTitle) {
+        filterOptions.stream()
+                .filter(option -> option.find(By.xpath(".//label")).getText().contains(filterOptionTitle))
```
Comment (toleksyn) [inline src/main/java/page_elements/filter_section/Filter.java]: chain with new line if more than 2 methods

File: src/main/java/page_elements/filter_section/Filter.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+                .limit(1)
+                .forEach(SelenideElement::click);
+    }
+}
+
+
```
Comment (toleksyn) [inline src/main/java/page_elements/filter_section/Filter.java]: remove

File: src/main/java/page_elements/filter_section/FiltersBlock.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import com.codeborne.selenide.ElementsCollection;
+import com.codeborne.selenide.SelenideElement;
+import lombok.Getter;
+import org.openqa.selenium.By;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Getter
+public class FiltersBlock {
```
Comment (toleksyn) [inline src/main/java/page_elements/filter_section/FiltersBlock.java]: try to have more detailed decomposition

File: src/main/java/page_elements/filter_section/FiltersBlock.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+                    filter.setFilterTitle(filterElement.find(By.xpath(".//button[@class='sidebar-block__toggle']")).getText());
+                    filter.setFilterOptions(filterElement.findAll((By.xpath(".//a[@class='checkbox-filter__link']"))));
+                    availableFilters.add(filter);
+                });
+
+    }
+
+    public void selectFilter(String filterTitle, String filterOptionTitle) {
+        availableFilters.stream()
+                .filter(s -> s.getFilterTitle().contains(filterTitle))
```
Comment (toleksyn) [inline src/main/java/page_elements/filter_section/FiltersBlock.java]: rename s

File: src/main/java/page_elements/items_grid/ItemTile.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+public class ItemTile {
+    private SelenideElement itemLink;
+    private String itemPrice;
+    private String rate;
+    private boolean isAvailable;
+    private SelenideElement shoppingCartButton;
+    private SelenideElement addToWishListButton;
+    private SelenideElement addToComparisonButton;
+    private SelenideElement reviewsLink;
+
```
Comment (toleksyn) [inline src/main/java/page_elements/items_grid/ItemTile.java]: fix

File: src/main/java/page_elements/items_grid/ItemTile.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import com.codeborne.selenide.SelenideElement;
+import lombok.Data;
+import lombok.Getter;
+import lombok.Setter;
+import page_objects.ItemPage;
+import page_objects.ReviewsPage;
+
+@Getter
+@Setter
+public class ItemTile {
```
Comment (toleksyn) [inline src/main/java/page_elements/items_grid/ItemTile.java]: be specific - ProductTile

File: src/main/java/page_elements/items_grid/ItemTile.java (hunk 7 above)
Comment (toleksyn) [reply src/main/java/page_elements/items_grid/ItemTile.java]: add behaviour

File: src/main/java/page_elements/items_grid/ItemsGrid.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.codeborne.selenide.Selenide.$x;
+
+@Getter
+public class ItemsGrid {
+    private List<ItemTile> itemTiles = new ArrayList<>();
+
+    public ItemsGrid(ElementsCollection collection) {
```
Comment (toleksyn) [inline src/main/java/page_elements/items_grid/ItemsGrid.java]: move to method getTiles(limitNumber)

File: src/main/java/page_elements/items_grid/ItemsGrid.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Getter;
+import org.openqa.selenium.By;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.codeborne.selenide.Selenide.$x;
+
+@Getter
+public class ItemsGrid {
```
Comment (toleksyn) [inline src/main/java/page_elements/items_grid/ItemsGrid.java]: ProductsGrid

File: src/main/java/page_elements/items_grid/ItemsGrid.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    private List<ItemTile> itemTiles = new ArrayList<>();
+
+    public ItemsGrid(ElementsCollection collection) {
+        collection.stream()
+                .forEach(itemTileElement -> {
+                    ItemTile tile = new ItemTile();
+                    itemTileElement.shouldBe(Condition.enabled);
+                    tile.setItemPrice(itemTileElement.find(By.xpath(".//span[@class='goods-tile__price-value']")).getText());
```
Comment (toleksyn) [inline src/main/java/page_elements/items_grid/ItemsGrid.java]: avoid inline method calls, extract to variables

File: src/main/java/page_objects/BasePage.java [hunk 12]
```diff
@@ -1,42 +1,44 @@
...
-import static com.codeborne.selenide.Selenide.sleep;
 
-public class BasePage {
-    protected SelenideElement topUserAccountLink = $("a.header-topline__user-link");
+@Getter
+public abstract class BasePage {
+
+    private SelenideElement topUserAccountLink = $("a.header-topline__user-link");
 
     @Step("Log in as user")
```
Comment (toleksyn) [inline src/main/java/page_objects/BasePage.java]: extract to header

File: src/main/java/page_objects/BasePage.java [hunk 13]
```diff
@@ -1,42 +1,44 @@
...
         $("#auth_email").sendKeys(email);
         $("#auth_pass").sendKeys(password);
         $(".auth-modal__remember-checkbox").click();
-        $("div.auth-modal__form-bottom > button ").shouldBe(Condition.enabled).click();
+        $("div.auth-modal__form-bottom > button ").shouldBe(enabled).click();
+        topUserAccountLink.shouldHave(attribute("href", "https://my.rozetka.com.ua/profile/personal-information"));
         return new HomePage();
     }
 
     public String getNameOfLoggedInUser() {
```
Comment (toleksyn) [inline src/main/java/page_objects/BasePage.java]: getLoggedInUserName()

File: src/main/java/page_objects/CheckoutDialogPage.java [hunk 14]
```diff
@@ -0,0 +1,4 @@
+package page_objects;
+
+public class CheckoutDialogPage {
+}
```
Comment (toleksyn) [inline src/main/java/page_objects/CheckoutDialogPage.java]: // extension purpose only
add comment

File: src/main/java/page_objects/HomePage.java [hunk 15]
```diff
@@ -1,15 +1,37 @@
...
+        return category;
+    }
+
     @Step("Open home page")
-    public static final HomePage openPage() {
+    public static HomePage openHomePage() {
         open("https://rozetka.com.ua/");
+        //sleep is very bad practice
+        //but without it I constantly have StaleElementExeption error
+        Selenide.sleep(2000);
```
Comment (toleksyn) [inline src/main/java/page_objects/HomePage.java]: avoid, element collection doesn't wait, use .shouldBe()

File: src/main/java/page_objects/ItemPage.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+import com.codeborne.selenide.Condition;
+import com.codeborne.selenide.Selenide;
+import com.codeborne.selenide.SelenideElement;
+import org.openqa.selenium.By;
+
+import static com.codeborne.selenide.Selenide.$;
+import static com.codeborne.selenide.Selenide.$x;
+
+public class ItemPage {
+    private SelenideElement reviewLinks = $x("//a[@class='product__rating-reviews']");
```
Comment (toleksyn) [inline src/main/java/page_objects/ItemPage.java]: singular

File: src/main/java/page_objects/ItemPage.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+    public int getItemPrice() {
+        //used regex which replaces all non-digit characters with empty string
```
Comment (toleksyn) [inline src/main/java/page_objects/ItemPage.java]: no need to comment self explanatory code

File: src/main/java/page_objects/ItemPage.java [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+        return $x("//app-rating-stars").find(By.cssSelector("svg")).getAttribute("aria-label");
+    }
+
+    public int getAmountOfReviews() {
+        return Integer.parseInt(reviewLinks.getText().replaceAll("\\D", ""));
+    }
+
+    public ItemPage openReviewsTab() {
+        reviewLinks.click();
+        return this;
```
Comment (toleksyn) [inline src/main/java/page_objects/ItemPage.java]: create an object

File: src/main/java/page_objects/ItemPage.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    public String getHeaderOfReviewsSection() {
```
Comment (toleksyn) [inline src/main/java/page_objects/ItemPage.java]: getReviewTitle()

File: src/main/java/page_objects/ListItemsPage.java [hunk 20]
```diff
@@ -0,0 +1,95 @@
+package page_objects;
+
+
```
Comment (toleksyn) [inline src/main/java/page_objects/ListItemsPage.java]: remove

File: src/main/java/page_objects/ListItemsPage.java [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+import static com.codeborne.selenide.Condition.visible;
+import static com.codeborne.selenide.Selenide.*;
+
+public class ListItemsPage {
+    protected ItemsGrid itemsGrid;
+    protected FiltersBlock filterSidebar;
+
+    public ListItemsPage() {
+        Selenide.sleep(3000);
```
Comment (toleksyn) [inline src/main/java/page_objects/ListItemsPage.java]: remove

File: src/main/java/page_objects/ListItemsPage.java [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+    public ListItemsPage selectFilter(String filterTitle, String filterOptionTitle) {
+        filterSidebar.selectFilter(filterTitle, filterOptionTitle);
+        return new ListItemsPage();
+    }
+
+    public int getAmountOfFeedbacks(int index) {
+        return Integer.parseInt(getItemTile(index).getReviewsLink().getText().replaceAll("\\D", ""));
+    }
+
+
```
Comment (toleksyn) [inline src/main/java/page_objects/ListItemsPage.java]: remove,
move behaviour to tile object

File: src/main/java/page_objects/ReviewsPage.java [hunk 23]
```diff
@@ -0,0 +1,4 @@
+package page_objects;
+
+public class ReviewsPage {
+}
```
Comment (toleksyn) [inline src/main/java/page_objects/ReviewsPage.java]: !!

File: src/main/java/page_objects/SearchResultsPage.java [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+import com.codeborne.selenide.ElementsCollection;
+import com.codeborne.selenide.SelenideElement;
+import org.openqa.selenium.By;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.codeborne.selenide.Selenide.*;
+
+
```
Comment (toleksyn) [inline src/main/java/page_objects/SearchResultsPage.java]: !!

File: src/main/java/page_objects/SearchResultsPage.java [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+                .limit(amountOfItems)
+                .collect(Collectors.toList());
+
+    }
+
+    public String getSearchResultsTitle() {
+        return $x("//h1[@class='catalog-heading']").getText().replaceAll("[«,»]", "");
+    }
+
+}
```
Comment (toleksyn) [inline src/main/java/page_objects/SearchResultsPage.java]: re-do into objects

File: src/main/java/page_objects/WishListPage.java [hunk 26]
```diff
@@ -0,0 +1,4 @@
+package page_objects;
+
+public class WishListPage {
+}
```
Comment (toleksyn) [inline src/main/java/page_objects/WishListPage.java]: !!

File: src/test/java/tests/FilterTest.java [hunk 27]
```diff
@@ -0,0 +1,65 @@
+package tests;
```
Comment (toleksyn) [inline src/test/java/tests/FilterTest.java]: com.softserveinc.atqc for all packages

File: src/test/java/tests/FilterTest.java [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+public class FilterTest {
+
+    @BeforeMethod
+    public void setup() {
+        Selenide.clearBrowserCookies();
+        Configuration.startMaximized = true;
+    }
+
+    @Test
+
```
Comment (toleksyn) [inline src/test/java/tests/FilterTest.java]: !!

File: src/main/java/page_objects/HomePage.java [hunk 29]
```diff
@@ -1,15 +1,37 @@
...
+                .filter(element -> element.getAttribute("href").contains(categoryTitle))
+                .findFirst()
+                .get();
+
+        return category;
+    }
+
     @Step("Open home page")
-    public static final HomePage openPage() {
+    public static HomePage openHomePage() {
```
Comment (toleksyn) [inline src/main/java/page_objects/HomePage.java]: no static methods in PO

File: src/test/java/tests/FilterTest.java [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+    public void verifyThatFilteredItemHasCorrectPriceAndTitle() {
+
+        val listItemsPage = HomePage.openHomePage()
```
Comment (toleksyn) [inline src/test/java/tests/FilterTest.java]: fix by inlining selenide elements to getter or methods

Comment (toleksyn) [reply src/test/java/tests/FilterTest.java]: new HomePage().open()

File: src/test/java/tests/FilterTest.java [hunk 31]
```diff
@@ -0,0 +1,65 @@
...
+        String actualTitle = itemPage.getItemTitle();
+        int actualPrice = itemPage.getItemPrice();
+        String actualRate = itemPage.getItemRate();
+//        Selenide.sleep(20000);
+
+        Assert.assertEquals(actualPrice, expectedPrice);
+        Assert.assertEquals(actualTitle, expectedTitle);
+
+        //Assert.assertEquals(actualRate,expectedRate);
+    }
```
Comment (toleksyn) [inline src/test/java/tests/FilterTest.java]: val

File: src/test/java/tests/FilterTest.java [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+        val itemPage = listItemsPage.openItemsPage(0);
+        String actualTitle = itemPage.getItemTitle();
+        int actualPrice = itemPage.getItemPrice();
+        String actualRate = itemPage.getItemRate();
+//        Selenide.sleep(20000);
+
+        Assert.assertEquals(actualPrice, expectedPrice);
+        Assert.assertEquals(actualTitle, expectedTitle);
+
+        //Assert.assertEquals(actualRate,expectedRate);
```
Comment (toleksyn) [inline src/test/java/tests/FilterTest.java]: add comment

File: src/test/java/tests/FilterTest.java [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+
+        Assert.assertEquals(amountOfFeedbacksonItemPage, amountOfFeedbacksOnSearchResultPage);
+
+        val headerOfReviewSection = itemPage.openReviewsTab().getHeaderOfReviewsSection();
+        Assert.assertEquals(headerOfReviewSection, "Отзывы покупателей о " + itemPage.getItemTitle() + " " + itemPage.getAmountOfReviews());
+
+
+    }
+
+}
```
Comment (toleksyn) [inline src/test/java/tests/FilterTest.java]: !!

File: src/test/java/tests/FilterTest.java [hunk 34]
```diff
@@ -0,0 +1,65 @@
...
+        String expectedRate = listItemsPage.getItemRate(0);
+
+        val itemPage = listItemsPage.openItemsPage(0);
+        String actualTitle = itemPage.getItemTitle();
+        int actualPrice = itemPage.getItemPrice();
+        String actualRate = itemPage.getItemRate();
+//        Selenide.sleep(20000);
+
+        Assert.assertEquals(actualPrice, expectedPrice);
+        Assert.assertEquals(actualTitle, expectedTitle);
```
Comment (toleksyn) [inline src/test/java/tests/FilterTest.java]: stat imp

File: src/test/java/tests/LoginTest.java [hunk 35]
```diff
@@ -0,0 +1,33 @@
...
+                .logIn("ezbooksforme@gmail.com", "Pa55word");
+        val nameOfLoggedInUser = homePage.getNameOfLoggedInUser();
+        assertEquals(nameOfLoggedInUser, expectedUserName);
+
+        homePage.navigateToUserAccount().logOutUsingLinkInProfile();
+        Assert.assertEquals(homePage.getNameOfLoggedInUser(), loginLinkLabel);
+
+    }
+
+
```
Comment (toleksyn) [inline src/test/java/tests/LoginTest.java]: !!!

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 1 above)
Comment (sovchenko) [reply Jenkinsfile]: done

File: src/main/java/page_elements/filter_section/Filter.java (hunk 2 above)
Comment (sovchenko) [reply src/main/java/page_elements/filter_section/Filter.java]: done

File: src/main/java/page_elements/filter_section/FiltersBlock.java (hunk 5 above)
Comment (sovchenko) [reply src/main/java/page_elements/filter_section/FiltersBlock.java]: done

File: src/main/java/page_elements/filter_section/Filter.java (hunk 4 above)
Comment (sovchenko) [reply src/main/java/page_elements/filter_section/Filter.java]: done

File: src/main/java/page_elements/filter_section/Filter.java (hunk 3 above)
Comment (sovchenko) [reply src/main/java/page_elements/filter_section/Filter.java]: done

File: src/test/java/tests/FilterTest.java (hunk 31 above)
Comment (sovchenko) [reply src/test/java/tests/FilterTest.java]: done

File: src/test/java/tests/FilterTest.java (hunk 32 above)
Comment (sovchenko) [reply src/test/java/tests/FilterTest.java]: assertion was removed

File: src/test/java/tests/FilterTest.java (hunk 33 above)
Comment (sovchenko) [reply src/test/java/tests/FilterTest.java]: done

File: src/test/java/tests/FilterTest.java (hunk 34 above)
Comment (sovchenko) [reply src/test/java/tests/FilterTest.java]: done

File: src/test/java/tests/LoginTest.java (hunk 35 above)
Comment (sovchenko) [reply src/test/java/tests/LoginTest.java]: done

**Comment by sovchenko** [comment]
concerns from all comments were resolved

**Pull request closed without merging**